common/quotas/clocked_rate_limiter.go

```diff
@@ -0,0 +1,154 @@
+// the time determined by the clock.
+type ClockedRateLimiter struct {
+	rateLimiter *rate.Limiter
+	clock       clockwork.Clock
+}
+
+var (
+	ErrRateLimiterWaitInterrupted                       = errors.New("rate limiter wait interrupted")
+	ErrRateLimiterReservationCannotBeMade               = errors.New("rate limiter reservation cannot be made due to insufficient quota")
+	ErrRateLimiterReservationWouldExceedContextDeadline = errors.New("rate limiter reservation would exceed context deadline")
+)
+
+func NewClockedRateLimiter(rateLimiter *rate.Limiter, clock clockwork.Clock) ClockedRateLimiter {
+	return ClockedRateLimiter{
+		rateLimiter: rateLimiter,
+		clock:       clock,
+	}
+}
+
+func (l ClockedRateLimiter) Allow() bool {
+	return l.AllowN(l.clock.Now(), 1)
+}
+
+func (l ClockedRateLimiter) AllowN(now time.Time, token int) bool {
+	return l.rateLimiter.AllowN(now, token)
+}
+
+// ClockedReservation wraps a rate.Reservation with a clockwork.Clock. It is used to ensure that the reservation
+// respects the time determined by the clock.
+type ClockedReservation struct {
+	reservation *rate.Reservation
+	clock       clockwork.Clock
+}
+
+func (r ClockedReservation) OK() bool {
+	return r.reservation.OK()
+}
+
+func (r ClockedReservation) Delay() time.Duration {
+	return r.DelayFrom(r.clock.Now())
+}
+
+func (r ClockedReservation) DelayFrom(t time.Time) time.Duration {
+	return r.reservation.DelayFrom(t)
+}
+
+func (r ClockedReservation) Cancel() {
+	r.CancelAt(r.clock.Now())
+}
+
+func (r ClockedReservation) CancelAt(t time.Time) {
+	r.reservation.CancelAt(t)
+}
+
+func (l ClockedRateLimiter) Reserve() ClockedReservation {
+	return l.ReserveN(l.clock.Now(), 1)
+}
+
+func (l ClockedRateLimiter) ReserveN(now time.Time, token int) ClockedReservation {
+	reservation := l.rateLimiter.ReserveN(now, token)
+	return ClockedReservation{reservation, l.clock}
+}
+
+func (l ClockedRateLimiter) Wait(ctx context.Context) error {
+	return l.WaitN(ctx, 1)
+}
+
+// WaitN is the only method that is different from rate.Limiter. We need to fully reimplement this method because
+// the original method uses time.Now(), and does not allow us to pass in a time.Time. Fortunately, it can be built on
+// top of ReserveN. However, there are some optimizations that we can make.
+func (l ClockedRateLimiter) WaitN(ctx context.Context, token int) error {
+	reservation := ClockedReservation{l.rateLimiter.ReserveN(l.clock.Now(), token), l.clock}
+	if !reservation.OK() {
+		return fmt.Errorf(
+			"%w: reservation would delay for %v",
+			ErrRateLimiterReservationCannotBeMade,
+			reservation.Delay(),
+		)
+	}
+
+	waitDuration := reservation.Delay()
+
+	// Optimization: if the waitDuration is 0, we don't need to start a timer.
+	if waitDuration <= 0 {
+		return nil
+	}
+
+	// Optimization: if the waitDuration is longer than the context deadline, we can immediately return an error.
```

Comment:
just to confirm, this is also the existing behavior of Wait() right?

Reply:
Yes--I found this in the source code for it actually: https://cs.opensource.google/go/x/time/+/refs/tags/v0.3.0:rate/rate.go;l=273